Add encryption to azure_rm_galleryimageversion

ephracis opened the pull request:
##### SUMMARY
Add the parameter `encryption` to `target_regions` of the `azure_rm_galleryimageversion` module.

Fixes #1290

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
azure_rm_galleryimageversion

##### ADDITIONAL INFORMATION
This makes it possible to create encrypted images.

Had to bump API version to get access to the new fields.

ephracis committed:
Add encryption to azure_rm_galleryimageversion

Add the parameter `encryption` to `target_regions` of the
`azure_rm_galleryimageversion` module, making it possible to create
encrypted images.

Fixes #1290

Fred-sun commented:
@ephracis Please help fix the error described below, the new parameters also need to be supported by the test case! Thank you very much!
```e.g.
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: missing-suboption-docs: Argument 'confidential_vm_encryption_type' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile has sub-options but documentation does not define it
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: missing-suboption-docs: Argument 'securityProfile' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image has sub-options but documentation does not define it
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: nonexistent-parameter-documented: Argument 'security_profile' found in publishing_profile -> target_regions -> encryption -> os_disk_image is listed in DOCUMENTATION.options, but not accepted by the module argument_spec
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-list-no-elements: Argument 'data_disk_images' in argument_spec found in publishing_profile -> target_regions -> encryption defines type as list but elements is not defined
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-list-no-elements: DOCUMENTATION.options.publishing_profile.suboptions.target_regions.suboptions.encryption.suboptions.data_disk_images: Argument defines type as list but elements is not defined for dictionary value @ data['options']['publishing_profile']['suboptions']['target_regions']['suboptions']['encryption']['suboptions']['data_disk_images']. Got {'description': ['A list of encryption specifications for data disk images.'], 'type': 'list', 'suboptions': {'disk_encryption_set_id': {'description': ['A relative URI containing the resource ID of the disk encryption set.'], 'type': 'str'}, 'lun': {'description': ['This property specifies the logical unit number of the data disk.', 'This value is used to identify data disks within the Virtual Machine and therefore must be unique for each data disk attached to the Virtual Machine.']}}}
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-type-not-in-doc: Argument 'confidential_vm_encryption_type' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-type-not-in-doc: Argument 'encrypted_vm_guest_state_only_with_pmk' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile -> confidential_vm_encryption_type defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-type-not-in-doc: Argument 'encrypted_with_cmk' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile -> confidential_vm_encryption_type defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-type-not-in-doc: Argument 'encrypted_with_pmk' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile -> confidential_vm_encryption_type defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-type-not-in-doc: Argument 'lun' in argument_spec found in publishing_profile -> target_regions -> encryption -> data_disk_images defines type as 'int' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-type-not-in-doc: Argument 'secure_vm_disk_encryption_set_id' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: parameter-type-not-in-doc: Argument 'securityProfile' in argument_spec found in publishing_profile -> target_regions -> encryption -> os_disk_image defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: undocumented-parameter: Argument 'confidential_vm_encryption_type' found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: undocumented-parameter: Argument 'encrypted_vm_guest_state_only_with_pmk' found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile -> confidential_vm_encryption_type is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: undocumented-parameter: Argument 'encrypted_with_cmk' found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile -> confidential_vm_encryption_type is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: undocumented-parameter: Argument 'encrypted_with_pmk' found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile -> confidential_vm_encryption_type is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: undocumented-parameter: Argument 'secure_vm_disk_encryption_set_id' found in publishing_profile -> target_regions -> encryption -> os_disk_image -> securityProfile is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_galleryimageversion.py:0:0: undocumented-parameter: Argument 'securityProfile' found in publishing_profile -> target_regions -> encryption -> os_disk_image is listed in the argument_spec, but not documented in the module documentation

ephracis commented:
Sure thing! I'll start working on it right away.

ephracis committed:
Fix documentation

ephracis committed:
Add disk encryption to gallery image version tests

Add the usage of disk encryption to all regions in the tests of the
module azure_rm_galleryimageversion.

This required a fix, moving the deprecated managed_image to
storage_profile.source_image instead.

ephracis commented:
I have now fixed the issues in the module specification, and added integrations tests.

xuzhang3 commented on tests/integration/targets/azure_rm_gallery/lookup_plugins/azure_service_principal_attribute.py:
```diff
@@ -0,0 +1,92 @@
+# (c) 2018 Yunge Zhu, <yungez@microsoft.com>
```
@ephracis you are the author.

Fred-sun replied:
> @ephracis you are the author.

@xuzhang3 This file was copied from another location and no changes are required to the author. Thank you!

ephracis replied:
Yes, that's true. It would actually be better to put the file somewhere where it could be shared by both integration tests but this was my first attempt at it.

If you have any suggestions on where to put the file and then share it, let me know and I can fix it as part of this PR.

I'm actually wondering why this is not part of the collection itself, and only part of the integration tests. Seems like a good plugin to have. But perhaps there's a reason you don't want it shipped.

Fred-sun replied:
@ephracis Yes, we are considering it as a plugin that can be applied to the entire repo. Thanks!
